vibrato/src/dictionary/lexicon/builder.rs

```diff
@@ -39,27 +30,88 @@ impl Lexicon {
-
-        Ok(RawWordEntry {
-            surface,
-            param: WordParam::new(left_id, right_id, word_cost),
-            feature,
-        })
+        let mut rdr = csv_core::Reader::new();
+        let mut features_bytes = bytes;
+        let mut record_bytes = bytes;
+        let mut field_cnt: usize = 0;
+        let mut features_len = 0;
+        let mut record_end_pos = 0;
+        let mut output = [0; 4096];
+
+        let mut surface = String::new();
+        let mut left_id = 0;
+        let mut right_id = 0;
+        let mut word_cost = 0;
+
+        loop {
+            let (result, nin, nout) = rdr.read_field(bytes, &mut output);
+            let record_end = match result {
+                ReadFieldResult::InputEmpty => {
+                    features_len += nin + 1;
+                    record_end_pos += nin;
+                    true
+                }
+                ReadFieldResult::OutputFull => {
+                    return Err(VibratoError::invalid_format(name, "Field too large"))
+                }
+                ReadFieldResult::Field { record_end } => {
+                    match field_cnt {
+                        0 => {
+                            surface = std::str::from_utf8(&output[..nout])?.to_string();
+                            record_bytes = bytes;
+                        }
+                        1 => {
+                            left_id = std::str::from_utf8(&output[..nout])?.parse()?;
+                        }
+                        2 => {
+                            right_id = std::str::from_utf8(&output[..nout])?.parse()?;
+                        }
+                        3 => {
+                            word_cost = std::str::from_utf8(&output[..nout])?.parse()?;
+                            features_bytes = &bytes[nin..];
+                            features_len = 0;
+                        }
+                        _ => {
+                            features_len += nin;
+                        }
+                    }
+                    record_end_pos += nin;
+                    record_end
+                }
+                ReadFieldResult::End => break,
+            };
+            if record_end {
+                if field_cnt == 0 && nin == 0 {
+                    continue;
+                }
+                if field_cnt <= 3 {
+                    let msg = format!(
+                        "A csv row of lexicon must have five items at least, {:?}",
+                        std::str::from_utf8(&record_bytes[..record_end_pos])?,
+                    );
+                    return Err(VibratoError::invalid_format(name, msg));
+                }
+                let feature = std::str::from_utf8(&features_bytes[..features_len - 1])?;
+                entries.push(RawWordEntry {
+                    surface,
+                    param: WordParam::new(left_id, right_id, word_cost),
+                    feature,
+                });
```

Comment:
```suggestion
                if !surface.is_empty() {
                    let feature = std::str::from_utf8(&features_bytes[..features_len - 1])?;
                    entries.push(RawWordEntry {
                        surface,
                        param: WordParam::new(left_id, right_id, word_cost),
                        feature,
                    });
                }
```
Could you add the IF condition to ignore an empty surface since it is not handled in the analysis?
(Unfortunately, unidic-cwj-3.1.0 includes such an entry. Of course, an empty entry can be defined as an invalid argument. But, the current version ignores such entries internally to use unidic-cwj easily.)

Reply:
Fixed.
In the CSV specification, the number of iterations does not indicate the row number, so I modified the error message to output the entry string instead of the row number.